File: schemakernel/exceptions.py
from __future__ import annotations

from typing import Any


class SchemakernelError(Exception):
    pass


# Planner layer
class PlannerError(SchemakernelError):
    pass


class PlannerResponseError(PlannerError):
    pass


class PlannerRetryExhausted(PlannerError):
    pass


# Validation layer
class ValidationError(SchemakernelError):
    pass


class FieldValidationError(ValidationError):
    def __init__(self, key: str, reason: str, value: Any = None) -> None:
        self.key = key
        self.reason = reason
        self.value = value
        super().__init__(f"Field '{key}': {reason}")


class PolicyViolationError(ValidationError):
    pass


class PlannerOutputRejected(ValidationError):
    pass


# Workflow layer
class WorkflowError(SchemakernelError):
    pass


class TurnLimitExceeded(WorkflowError):
    pass


class SessionNotFound(WorkflowError):
    pass


class InvalidTransition(WorkflowError):
    pass

File: js-sdk/src/exceptions.ts
export class SchemaKernelError extends Error {
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
  }
}

// Planner layer
export class PlannerError extends SchemaKernelError {}

export class PlannerResponseError extends PlannerError {}

export class PlannerRetryExhausted extends PlannerError {}

// Validation layer
export class ValidationError extends SchemaKernelError {}

export class FieldValidationError extends ValidationError {
  constructor(public key: string, public reason: string, public value?: any) {
    super(`Field '${key}': ${reason}`);
    this.name = "FieldValidationError";
  }
}

export class PolicyViolationError extends ValidationError {}

export class PlannerOutputRejected extends ValidationError {}

// Workflow layer
export class WorkflowError extends SchemaKernelError {}

export class TurnLimitExceeded extends WorkflowError {}

export class SessionNotFound extends WorkflowError {}

export class InvalidTransition extends WorkflowError {}
